quinn/src/lib.rs

```
@@ -481,137 +414,91 @@ impl EndpointInner {
-            if sent >= IO_LOOP_BOUND {
+            if blocked {
                 return Ok(true);
             }
         }
-        Ok(false)
+        Ok(blocked)
     }
 
-    fn drive_timers(&mut self, now: Instant) -> Result<bool, io::Error> {
+    fn handle_events(&mut self) -> Result<bool, io::Error> {
+        use EndpointEvent::*;
         let mut keep_going = false;
         loop {
-            match self.timers.poll() {
-                Ok(Async::Ready(Some(Some((ch, timer))))) => {
-                    trace!(self.log, "timeout"; "timer" => ?timer);
-                    self.inner.timeout(now, ch, timer);
-                    if timer == quinn::Timer::Close {
-                        // Connection drained
-                        if let hash_map::Entry::Occupied(mut p) = self.pending.entry(ch) {
-                            if let Some(x) = p.get_mut().closing.take() {
-                                let _ = x.send(());
+            match self.events.poll() {
+                Ok(Async::Ready(Some((ch, event)))) => {
+                    match event {
+                        Connected => {
+                            self.connecting.remove(&ch);
+                        }
+                        Proto(e) => {
+                            if let quinn::EndpointEvent::Closed { .. } = &e {
+                                self.connections.remove(&ch);
                             }
-                            if p.get().dropped {
-                                p.remove();
-                            } else {
-                                p.get_mut().drained = true;
+                            if let Some(event) = self.inner.handle_event(ch, e) {
+                                if let hash_map::Entry::Occupied(mut entry) =
+                                    self.connections.entry(ch)
+                                {
+                                    entry.get_mut().unbounded_send(event).unwrap();
+                                }
```

What are the circumstances under which we expect to receive an event for an unknown connection? If none, we probably shouldn't silently ignore the inconsistent state.

---

Made it noisy!